src/Dfe.PlanTech.Application/Response/Commands/ProcessCheckAnswerDtoCommand.cs

```diff
@@ -40,97 +33,39 @@ public async Task CalculateMaturityAsync(int submissionId)
+        private async Task<CheckAnswerDto> _RemoveDetachedQuestions(List<QuestionWithAnswer> questionWithAnswerList, string sectionId)
         {
-            var answer = await _GetResponseAnswer(answerId);
-            string answerContentfulRef = answer?.ContentfulRef ?? throw new NullReferenceException(nameof(answer.ContentfulRef));
-            string answerText = answer?.AnswerText ?? throw new NullReferenceException(nameof(answerText));
+            if (questionWithAnswerList == null) throw new ArgumentNullException(nameof(questionWithAnswerList));
+            if (sectionId == null) throw new ArgumentNullException(nameof(sectionId));
 
-            return new QuestionWithAnswer()
-            {
-                QuestionRef = questionRef,
-                QuestionText = questionText,
-                AnswerRef = answerContentfulRef,
-                AnswerText = answerText
-            };
-        }
+            CheckAnswerDto checkAnswerDto = new CheckAnswerDto();
 
-        private async Task<CheckAnswerDto> _RemoveDetachedQuestions(CheckAnswerDto checkAnswerDto)
-        {
-            int questionAnswerListCount = checkAnswerDto.QuestionAnswerList.Count;
-            if (questionAnswerListCount <= 1) return checkAnswerDto;
+            Dictionary<string, QuestionWithAnswer> questionWithAnswerMap = questionWithAnswerList
+                            .Select((questionWithAnswer, index) => new KeyValuePair<string, QuestionWithAnswer>(questionWithAnswer.QuestionRef, questionWithAnswerList[index]))
+                            .ToDictionary(x => x.Key, x => x.Value);
 
-            Dictionary<string, bool> isDetachedMap = new Dictionary<string, bool>();
+            Section section = await _GetSection(sectionId) ?? throw new NullReferenceException(nameof(section));
 
-            for (int i = 0; i < questionAnswerListCount; i++) isDetachedMap.Add(checkAnswerDto.QuestionAnswerList[i].QuestionRef, i != 0);
+            Question? node = section.Questions.Where(question => question.Sys.Id.Equals(section.FirstQuestionId)).First();
 
-            foreach (QuestionWithAnswer questionWithAnswer in checkAnswerDto.QuestionAnswerList)
+            while (node != null)
             {
-                Domain.Questionnaire.Models.Answer answer = await _GetAnswer(questionWithAnswer.QuestionRef, questionWithAnswer.AnswerRef) ?? throw new NullReferenceException(nameof(answer));
-                string? nextQuestionId = answer.NextQuestion?.Sys.Id;
-                if (nextQuestionId == null) continue;
-                if (isDetachedMap.ContainsKey(nextQuestionId)) isDetachedMap[nextQuestionId] = false;
+                checkAnswerDto.QuestionAnswerList.Add(questionWithAnswerMap[node.Sys.Id]);
+                node = node.Answers.Where(answer => answer.Sys.Id.Equals(questionWithAnswerMap[node.Sys.Id].AnswerRef)).First().NextQuestion;
```

Comment:
I think `.First()` will error if none found, so should probably use .FirstOrDefault()?.NextQuestion

Reply:
Resolved in: [b5ed27e](https://github.com/DFE-Digital/plan-technology-for-your-school/pull/151/commits/b5ed27eb4390c37e305a6475929859d25e9c935c)